### PR TITLE
Integrate views with mixins

### PR DESCRIPTION
**Pull request by nahueespinosa**
### Proposed changes

This integrates `random_intersperse` and `sample` views into the core sampling mixins. It also creates a mixin subfolder to organize the library better.

Related to #279.

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [x] 🚀 Feature (change which adds functionality)
- [ ] 📚 Documentation (change which fixes or extends documentation)

### Checklist

- [x] Lint and unit tests (if any) pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### CODE REVIEW

**Comment by nahueespinosa** [comment]
So far, so good, in terms of micro-benchmark results:

Before integrating views: https://github.com/Ekumen-OS/beluga/commit/4f080fe91de5eb983f10ab64932bbff4825cb206

```
Running build/beluga/test/benchmark/benchmark_beluga
Run on (16 X 3300 MHz CPU s)
CPU Caches:
  L1 Data 32K (x8)
  L1 Instruction 32K (x8)
  L2 Unified 512K (x8)
  L3 Unified 16384K (x1)
Load Average: 3.25, 2.87, 2.27
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations
------------------------------------------------------------------------------------------------------
BM_FixedResample/128                                              7228 ns         7228 ns        96158 SampleSize=128
BM_FixedResample/256                                             15433 ns        15433 ns        45267 SampleSize=256
BM_FixedResample/512                                             32986 ns        32986 ns        21205 SampleSize=512
BM_FixedResample/1024                                            71258 ns        71257 ns         9809 SampleSize=1024
BM_FixedResample/2048                                           162614 ns       162603 ns         4298 SampleSize=2.048k
BM_FixedResample/4096                                           378110 ns       378094 ns         1863 SampleSize=4.096k
BM_FixedResample/8192                                           904698 ns       904678 ns          780 SampleSize=8.192k
BM_FixedResample/16384                                         1985326 ns      1985252 ns          345 SampleSize=16.384k
BM_FixedResample/32768                                         4377087 ns      4376741 ns          159 SampleSize=32.768k
BM_FixedResample/65536                                        10244166 ns     10243522 ns           68 SampleSize=65.536k
BM_FixedResample/131072                                       23804107 ns     23803229 ns           28 SampleSize=131.072k
BM_FixedResample/262144                                       56026337 ns     56024509 ns           13 SampleSize=262.144k
BM_FixedResample/524288                                      129831437 ns    129826646 ns            5 SampleSize=524.288k
BM_FixedResample/1000000                                     276883521 ns    276869696 ns            3 SampleSize=1000k
BM_FixedResample_BigO                                            13.60 NlgN      13.60 NlgN 
BM_FixedResample_RMS                                                10 %            10 %    
BM_AdaptiveResample/128                                          10086 ns        10086 ns        69808 Percentage=1 SampleSize=128
BM_AdaptiveResample/256                                          20285 ns        20285 ns        44841 Percentage=1 SampleSize=256
BM_AdaptiveResample/512                                          24172 ns        24169 ns        28126 Percentage=0.429688 SampleSize=220
BM_AdaptiveResample/1024                                         40348 ns        40348 ns        17940 Percentage=0.375977 SampleSize=385
BM_AdaptiveResample/2048                                         77479 ns        77478 ns         9180 Percentage=0.307129 SampleSize=629
BM_AdaptiveResample/4096                                        159120 ns       159117 ns         4390 Percentage=0.270996 SampleSize=1.11k
BM_AdaptiveResample/8192                                        331594 ns       331592 ns         2113 Percentage=0.247437 SampleSize=2.027k
BM_AdaptiveResample/16384                                       745020 ns       745000 ns          938 Percentage=0.231262 SampleSize=3.789k
BM_AdaptiveResample/32768                                      1525825 ns      1525827 ns          468 Percentage=0.217468 SampleSize=7.126k
BM_AdaptiveResample/65536                                      3964872 ns      3964791 ns          189 Percentage=0.210571 SampleSize=13.8k
BM_AdaptiveResample/131072                                     9546343 ns      9545964 ns           76 Percentage=0.204391 SampleSize=26.79k
BM_AdaptiveResample/262144                                    20954046 ns     20953001 ns           33 Percentage=0.201679 SampleSize=52.869k
BM_AdaptiveResample/524288                                    63313936 ns     63311889 ns            9 Percentage=0.198189 SampleSize=103.908k
BM_AdaptiveResample/1000000                                  125456181 ns    125454813 ns            5 Percentage=0.196877 SampleSize=196.877k
BM_AdaptiveResample_BigO                                          6.20 NlgN       6.20 NlgN 
BM_AdaptiveResample_RMS                                             17 %            17 %    
```

After integrating views (this PR):
```
Running build/beluga/test/benchmark/benchmark_beluga
Run on (16 X 3300 MHz CPU s)
CPU Caches:
  L1 Data 32K (x8)
  L1 Instruction 32K (x8)
  L2 Unified 512K (x8)
  L3 Unified 16384K (x1)
Load Average: 3.18, 2.49, 1.98
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations
------------------------------------------------------------------------------------------------------
BM_FixedResample/128                                              6320 ns         6320 ns       107663 SampleSize=128
BM_FixedResample/256                                             13733 ns        13732 ns        51380 SampleSize=256
BM_FixedResample/512                                             29730 ns        29730 ns        23433 SampleSize=512
BM_FixedResample/1024                                            65041 ns        65039 ns        10739 SampleSize=1024
BM_FixedResample/2048                                           140034 ns       140031 ns         4999 SampleSize=2.048k
BM_FixedResample/4096                                           326490 ns       326480 ns         2171 SampleSize=4.096k
BM_FixedResample/8192                                           776977 ns       776951 ns          900 SampleSize=8.192k
BM_FixedResample/16384                                         1747356 ns      1747330 ns          399 SampleSize=16.384k
BM_FixedResample/32768                                         3955691 ns      3955510 ns          177 SampleSize=32.768k
BM_FixedResample/65536                                         9487258 ns      9487039 ns           76 SampleSize=65.536k
BM_FixedResample/131072                                       21833048 ns     21832527 ns           32 SampleSize=131.072k
BM_FixedResample/262144                                       50612722 ns     50612354 ns           10 SampleSize=262.144k
BM_FixedResample/524288                                      115948992 ns    115947056 ns            6 SampleSize=524.288k
BM_FixedResample/1000000                                     232023200 ns    232021102 ns            3 SampleSize=1000k
BM_FixedResample_BigO                                            11.58 NlgN      11.58 NlgN 
BM_FixedResample_RMS                                                 6 %             6 %    
BM_AdaptiveResample/128                                           8555 ns         8555 ns        81859 Percentage=1 SampleSize=128
BM_AdaptiveResample/256                                          16302 ns        16302 ns        42119 Percentage=0.859375 SampleSize=220
BM_AdaptiveResample/512                                          27052 ns        27051 ns        25572 Percentage=0.630859 SampleSize=323
BM_AdaptiveResample/1024                                         45198 ns        45197 ns        15098 Percentage=0.504883 SampleSize=517
BM_AdaptiveResample/2048                                         84007 ns        84004 ns         7364 Percentage=0.423828 SampleSize=868
BM_AdaptiveResample/4096                                        172730 ns       172724 ns         3931 Percentage=0.366943 SampleSize=1.503k
BM_AdaptiveResample/8192                                        354104 ns       354102 ns         1994 Percentage=0.32959 SampleSize=2.7k
BM_AdaptiveResample/16384                                       803332 ns       803328 ns          868 Percentage=0.305298 SampleSize=5.002k
BM_AdaptiveResample/32768                                      1466143 ns      1466132 ns          469 Percentage=0.288635 SampleSize=9.458k
BM_AdaptiveResample/65536                                      3947176 ns      3947180 ns          178 Percentage=0.276962 SampleSize=18.151k
BM_AdaptiveResample/131072                                     8134551 ns      8134486 ns           86 Percentage=0.268875 SampleSize=35.242k
BM_AdaptiveResample/262144                                    19891934 ns     19891781 ns           37 Percentage=0.263245 SampleSize=69.008k
BM_AdaptiveResample/524288                                    44416100 ns     44414993 ns           16 Percentage=0.259363 SampleSize=135.981k
BM_AdaptiveResample/1000000                                  109723695 ns    109717529 ns            7 Percentage=0.256742 SampleSize=256.742k
BM_AdaptiveResample_BigO                                          5.23 NlgN       5.23 NlgN 
BM_AdaptiveResample_RMS                                             22 %            22 %   
```

File: beluga/test/benchmark/benchmark_take_while_kld.cpp [hunk 1]
```diff
@@ -93,22 +90,19 @@ void BM_AdaptiveResample(benchmark::State& state) {
...
-                     }) |
-                     ranges::views::common;
-    auto first = std::begin(new_container);
-    auto last = std::copy(std::begin(samples), std::end(samples), first);
-    state.counters["SampleSize"] = static_cast<double>(std::distance(first, last));
-    state.counters["Percentage"] = static_cast<double>(std::distance(first, last)) / static_cast<double>(max_samples);
+    auto samples =
+        container |              //
+        beluga::views::sample |  //
+        beluga::views::take_while_kld(beluga::spatial_hash<State>{}, min_samples, max_samples, kld_epsilon, kld_z);
```
Comment (hidmic) [inline beluga/test/benchmark/benchmark_take_while_kld.cpp:100]: @nahueespinosa meta: would it make sense for `beluga::views::take_while_kld` to use `beluga::spatial_hash<State>{}` if none is provided? IIUC we should be able to deduce `State` from `range_value_t<Range>` for the corresponding overload.

**Review by hidmic — APPROVED** [review]
LGTM. I don't fully grok `codecov`'s complaint.

File: beluga/test/benchmark/benchmark_take_while_kld.cpp (hunk 1 above)
Comment (nahueespinosa) [reply beluga/test/benchmark/benchmark_take_while_kld.cpp:100]: We can, but `spatial_hash` requires configuration parameters (i.e. resolution) that we would need to pass by value or provide sensible defaults. Both are reasonable. I'll sleep on it but if you have more thoughts on what API you'd like to see/use, that'd be great.

**Pull request merged**